src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.DependencyGraphItem.cs

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.DependencyResolver;
+using NuGet.LibraryModel;
+using NuGet.ProjectModel;
+
+namespace NuGet.Commands
+{
+    internal sealed partial class DependencyGraphResolver
+    {
+        /// <summary>
+        /// A class representing a declared dependency graph item to consider for the resolved graph.
+        /// </summary>
+        [DebuggerDisplay("{LibraryDependency}, DependencyIndex={LibraryDependencyIndex}, RangeIndex={LibraryRangeIndex}")]
+        public class DependencyGraphItem
+        {
+            /// <summary>
+            /// Gets or sets a <see cref="Task{TResult}" /> that returns a <see cref="GraphItem{TItem}" /> containing a <see cref="RemoteResolveResult" /> that represents the resolved graph item after looking it up in the configured feeds.
+            /// </summary>
+            public required Task<GraphItem<RemoteResolveResult>> FindLibraryTask { get; init; }
+
+            /// <summary>
+            /// Gets or sets a value indicating whether or not this dependency graph item is a transitively pinned dependency.
+            /// </summary>
+            public bool IsCentrallyPinnedTransitivePackage { get; init; }
+
+            /// <summary>
+            /// Gets or sets a value indicating whether or not this dependency graph item is a direct package reference from the root project.
+            /// </summary>
+            public bool IsRootPackageReference { get; init; }
+
+            /// <summary>
+            /// Gets or sets the <see cref="LibraryDependency" /> used to declare this dependency graph item.
+            /// </summary>
+            public required LibraryDependency LibraryDependency { get; init; }
+
+            /// <summary>
+            /// Gets or sets the <see cref="LibraryDependencyIndex" /> associated with this dependency graph item.
+            /// </summary>
+            public LibraryDependencyIndex LibraryDependencyIndex { get; init; } = LibraryDependencyIndex.Invalid;
+
+            /// <summary>
+            /// Gets or sets the <see cref="LibraryRangeIndex" /> associated with this dependency graph item.
+            /// </summary>
+            public LibraryRangeIndex LibraryRangeIndex { get; init; } = LibraryRangeIndex.Invalid;
+
+            /// <summary>
+            /// Gets or sets the <see cref="LibraryRangeIndex" /> of this dependency graph item's parent.
```

Comment:
Nit: Not sure if this is misleading as it's not an actual setter. 

```suggestion
            /// Gets or initializes the <see cref="LibraryRangeIndex" /> of this dependency graph item's parent.
```

Reply:
Done!